### PR TITLE
Add cluster-events filter

### PR DESCRIPTION
**Pull request by mareklibra**

### CODE REVIEW

**Comment by mareklibra** [comment]
---



---



---


---

**Comment by nirfarkas** [comment]
This looks great! 
Nits:
1. I'm not sure what's the Patternfly convention here, but I think it will be great to color highlight the searched term in the results (see quick mock), this helps the user understand the context.
2. When no events returned by the search criteria the text should be more accurate, maybe something like: "There are no events matching the requested criteria, try changing it to get results."
3. I would suggest changing the placeholder text inside the search field to be "Search" or "Free text search". I find the current one "Filter by content..." a bit confusing.

**Comment by mareklibra** [comment]
@nirfarkas your comments are addressed now

> BTW, does "Clear all filters" clears also the search?

Yes, it "resets" all filters, including the text search

**Review by jtomasek — APPROVED** [review]

File: src/components/ui/ClusterEventsList.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  );
+
+  const filteredEvents = filterEvenets(filters, events);
+
+  return (
+    <>
+      <ClusterEventsToolbar filters={filters} setFilters={setFilters} cluster={cluster} />
+      {filteredEvents.length === 0 && events.length > 0 ? (
+        <div>
+          There are no events matching the requested criteria, try changing it to get results.
```
Comment (jtomasek) [inline src/components/ui/ClusterEventsList.tsx:37]: Maybe we could use EmptyState here to handle the styling. With that we wouldn't need to add custom style and aligned to PF guidelines.

File: src/components/ui/EventsList.tsx [hunk 2]
```diff
@@ -39,13 +39,13 @@ const getLabelIcon = (severity: Event['severity']) => {
...
 
-type EventsListProps = {
+export type EventsListProps = {
   events: EventList;
 };
 
 const EventsList: React.FC<EventsListProps> = ({ events }) => {
   if (events.length === 0) {
-    return <div>No events</div>;
+    return <div>No events found.</div>;
```
Comment (jtomasek) [inline src/components/ui/EventsList.tsx:48]: EmptyState here as well.

**Review by jtomasek — CHANGES_REQUESTED** [review]

**Review by jtomasek — CHANGES_REQUESTED** [review]
I am getting following error when clicking the hosts filter dropdown on a cluster which does not have any hosts registered.
```
Uncaught Error: You can't have a focus-trap without at least one focusable element
    at getInitialFocusNode (index.js:198)
    at updateTabbableNodes (index.js:268)
    at Object.activate (index.js:72)
    at FocusTrap.componentDidMount (FocusTrap.tsx:43)
    at commitLifeCycles (react-dom.development.js:19847)
    at commitLayoutEffects (react-dom.development.js:22834)
    at HTMLUnknownElement.callCallback (react-dom.development.js:188)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:237)
    at invokeGuardedCallback (react-dom.development.js:292)
    at commitRootImpl (react-dom.development.js:22572)
    at unstable_runWithPriority (scheduler.development.js:653)
    at runWithPriority$1 (react-dom.development.js:11061)
    at commitRoot (react-dom.development.js:22412)
    at finishSyncRender (react-dom.development.js:21838)
    at performSyncWorkOnRoot (react-dom.development.js:21824)
    at react-dom.development.js:11111
    at unstable_runWithPriority (scheduler.development.js:653)
    at runWithPriority$1 (react-dom.development.js:11061)
    at flushSyncCallbackQueueImpl (react-dom.development.js:11106)
    at flushSyncCallbackQueue (react-dom.development.js:11094)
    at discreteUpdates$1 (react-dom.development.js:21924)
    at discreteUpdates (react-dom.development.js:1071)
    at dispatchDiscreteEvent (react-dom.development.js:4168)
getInitialFocusNode @ index.js:198
updateTabbableNodes @ index.js:268
activate @ index.js:72
componentDidMount @ FocusTrap.tsx:43
commitLifeCycles @ react-dom.development.js:19847
commitLayoutEffects @ react-dom.development.js:22834
callCallback @ react-dom.development.js:188
invokeGuardedCallbackDev @ react-dom.development.js:237
invokeGuardedCallback @ react-dom.development.js:292
commitRootImpl @ react-dom.development.js:22572
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11061
commitRoot @ react-dom.development.js:22412
finishSyncRender @ react-dom.development.js:21838
performSyncWorkOnRoot @ react-dom.development.js:21824
(anonymous) @ react-dom.development.js:11111
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11061
flushSyncCallbackQueueImpl @ react-dom.development.js:11106
flushSyncCallbackQueue @ react-dom.development.js:11094
discreteUpdates$1 @ react-dom.development.js:21924
discreteUpdates @ react-dom.development.js:1071
dispatchDiscreteEvent @ react-dom.development.js:4168
index.js:1 The above error occurred in the <FocusTrap> component:
    in FocusTrap (created by Context.Consumer)
    in SelectMenu (created by Select)
    in div (created by Select)
    in Select (created by ClusterEventsToolbar)
    in div (created by ToolbarItem)
    in ToolbarItem (created by ToolbarFilter)
    in ToolbarFilter (created by ClusterEventsToolbar)
    in div (created by Context.Consumer)
    in div (created by ToolbarContent)
    in ToolbarContent (created by ClusterEventsToolbar)
    in div (created by Toolbar)
    in Toolbar (created by ClusterEventsToolbar)
    in ClusterEventsToolbar (created by ClusterEventsList)
    in ClusterEventsList (created by EventListFetch)
    in EventListFetch (created by EventsModal)
    in div (created by EventsModal)
    in div (created by ModalBoxBody)
    in ModalBoxBody (created by ModalContent)
    in div (created by ModalBox)
    in ModalBox (created by ModalContent)
    in div (created by FocusTrap)
    in FocusTrap (created by ModalContent)
    in div (created by Backdrop)
    in Backdrop (created by ModalContent)
    in ModalContent (created by Modal)
    in Modal (created by EventsModal)
    in EventsModal (created by EventsModalButton)
    in EventsModalButton (created by Formik)
    in div (created by ToolbarGroupWithRef)
    in ToolbarGroupWithRef (created by ForwardRef)
    in ForwardRef (created by ToolbarSecondaryGroup)
    in ToolbarSecondaryGroup (created by Formik)
    in div (created by Context.Consumer)
    in div (created by ToolbarContent)
    in ToolbarContent (created by ClusterToolbar)
    in div (created by Toolbar)
    in Toolbar (created by ClusterToolbar)
    in section (created by PageSection)
    in PageSection (created by PageSection)
    in PageSection (created by ClusterToolbar)
    in ClusterToolbar (created by Formik)
    in Formik (created by ClusterConfigurationForm)
    in ClusterConfigurationForm (created by ClusterConfiguration)
    in ClusterConfiguration (created by ClusterPage)
    in AlertsContextProvider (created by ClusterPage)
    in ClusterPage (created by Context.Consumer)
    in Route (created by Router)
    in Switch (created by Router)
    in Router (at App.tsx:26)
    in main (created by Page)
    in div (created by Page)
    in Page (at App.tsx:19)
    in Router (at App.tsx:17)
    in Provider (at App.tsx:16)
    in App (at src/index.tsx:7)
```

**Comment by andybraren** [comment]
Thanks Marek!

For the Severity filter, could the behavior be similar to the OCP Web Console where all options are unchecked and the filter chip group doesn't appear by default? Here's an example from the Pods list where none of the filters are applied but all of the pods still appear:

image

**Comment by andybraren** [comment]
> > 1. I'm not sure what's the Patternfly convention here, but I think it will be great to color highlight the searched term in the results (see quick mock), this helps the user understand the context.
> 
> I do not remember seeing this in any related application but I agree that the user might be used to such behaviour from word-processing tools. Anyway, if this is not urgent, let's postpone implementation to a follow-up. Please file a jira task if you find useful tracking it.

Agreed, cool idea that I'm not sure PatternFly has a pattern for yet. @nirfarkas let's open up [a PatternFly issue](https://github.com/patternfly/patternfly/issues?q=is%3Aissue+is%3Aopen+) for discussion - I couldn't find any similar topic there.

> 2. When no events returned by the search criteria the text should be more accurate, maybe something like: "There are no events matching the requested criteria, try changing it to get results."

+1 to using the EmptyState as Jirka suggested to have the [native styling](https://www.patternfly.org/v4/design-guidelines/usage-and-behavior/empty-state). A secondary action button to "Reset filters" may also be cool.

Slight tweak of the current text for the empty state:

> [empty state icon] fa-search
> [title] No matching events
> [body] There are no events that match the current filters. Adjust or clear the filters to view events.
> [secondary button] Clear filters

> 3. I would suggest changing the placeholder text inside the search field to be "Search" or "Free text search". I find the current one "Filter by content..." a bit confusing.

It looks like OCM and other c.rh.c apps use `Filter by x...` which aligns with [PatternFly's guideline](https://www.patternfly.org/v4/design-guidelines/usage-and-behavior/filters#content). The OCP Web Console currently uses `Search by x...` but after a recent toolbar audit we opened [CONSOLE-2317](https://issues.redhat.com/browse/CONSOLE-2317) to change that wording to `Filter by x...` as well.

`Filter by content...` does sound a little weird here though. Maybe `Filter by text...` would be a little better for now?

I added this detail to our design review agenda Nir 🙂 - maybe we can chat with some PatternFly content folks about something like `Search` or `Search events` being clearer than any other `Filter by` option we can think of and follow up on this.

File: src/components/ui/ClusterEventsToolbar.tsx [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+              )),
+            ]}
+          </Select>
+        </ToolbarFilter>
+
+        <ToolbarFilter
+          chips={filters.severity}
+          deleteChip={onDeleteChip}
+          deleteChipGroup={onDeleteChipGroup}
+          categoryName="severity"
```
Comment (andybraren) [inline src/components/ui/ClusterEventsToolbar.tsx:169]:
```suggestion
          categoryName="Severity"
```

File: src/config/constants.ts [hunk 4]
```diff
@@ -125,3 +125,5 @@ export const HOST_VALIDATION_LABELS: { [key in Validation['id']]: string } = {
   'belongs-to-machine-cidr': 'Belongs to machine CIDR',
   'role-defined': 'Role',
 };
+
+export const EVENT_SEVERITIES: Event['severity'][] = ['info', 'warning', 'error', 'critical'];
```
Comment (andybraren) [inline src/config/constants.ts:129]:
```suggestion
export const EVENT_SEVERITIES: Event['severity'][] = ['Info', 'Warning', 'Error', 'Critical'];
```

I'm not sure this is the right place in the code to change this, but we should try to use sentence case throughout the UI.

**Review by andybraren — COMMENTED** [review]
Adding some code suggestions. I couldn't find where the text `Clear all filters` is defined, but it should be `Clear filters` per [PatternFly's filter behavior guideline](https://www.patternfly.org/v4/design-guidelines/usage-and-behavior/filters#clearing-filters) (way at the bottom).

File: src/components/ui/EventsList.tsx (hunk 2 above)
Comment (mareklibra) [reply src/components/ui/EventsList.tsx:48]: done

File: src/components/ui/ClusterEventsList.tsx (hunk 1 above)
Comment (mareklibra) [reply src/components/ui/ClusterEventsList.tsx:37]: done

**Comment by mareklibra** [comment]
- using EmptyState now
- rephrased to `Filter by text...` 
- severities are capitalized
- severities are empty by default but matching all. Badges added.

**Review by jtomasek — APPROVED** [review]
Looks great!

**Pull request merged**